x/dualstaking/keeper/delegate.go

```diff
@@ -165,106 +165,84 @@ func (k Keeper) decreaseDelegation(ctx sdk.Context, delegator, provider, chainID
+			stakeEntry.Stake, err = stakeEntry.Stake.SafeSub(amount)
 			if err != nil {
-				return utils.LavaFormatError("can't remove stake Entry after decreasing provider self delegation", err,
-					utils.Attribute{Key: "index", Value: index},
-					utils.Attribute{Key: "spec", Value: chainID},
-				)
+				return fmt.Errorf("invalid or insufficient funds: %w", err)
 			}
-			details["min_self_delegation"] = k.GetParams(ctx).MinSelfDelegation.String()
-			utils.LogLavaEvent(ctx, k.Logger(ctx), types.UnstakeFromUnbond, details, "unstaking provider due to unbond that lowered its stake below min self delegation")
-			unstakeHoldBlocks := k.epochstorageKeeper.GetUnstakeHoldBlocks(ctx, stakeEntry.Chain)
-			return k.epochstorageKeeper.AppendUnstakeEntry(ctx, stakeEntry, unstakeHoldBlocks)
-		} else if stakeEntry.EffectiveStake().LT(k.specKeeper.GetMinStake(ctx, chainID).Amount) {
-			details["min_spec_stake"] = k.specKeeper.GetMinStake(ctx, chainID).String()
-			utils.LogLavaEvent(ctx, k.Logger(ctx), types.FreezeFromUnbond, details, "freezing provider due to unbond that lowered its stake below min spec stake")
-			stakeEntry.Freeze()
 		}
 	} else {
-		stakeEntry.DelegateTotal, err = stakeEntry.DelegateTotal.SafeSub(amount)
+		if increase {
+			stakeEntry.DelegateTotal = stakeEntry.DelegateTotal.Add(amount)
+		} else {
+			stakeEntry.DelegateTotal, err = stakeEntry.DelegateTotal.SafeSub(amount)
+			if err != nil {
+				return fmt.Errorf("invalid or insufficient funds: %w", err)
+			}
+		}
+	}
+
+	details := map[string]string{
+		"provider":        stakeEntry.Address,
+		"chain_id":        stakeEntry.Chain,
+		"moniker":         stakeEntry.Moniker,
+		"stake":           stakeEntry.Stake.String(),
+		"effective_stake": stakeEntry.EffectiveStake().String() + stakeEntry.Stake.Denom,
+	}
+
+	if stakeEntry.Stake.IsLT(k.GetParams(ctx).MinSelfDelegation) {
+		err = k.epochstorageKeeper.RemoveStakeEntryCurrent(ctx, chainID, index)
 		if err != nil {
-			return fmt.Errorf("invalid or insufficient funds: %w", err)
+			return utils.LavaFormatError("can't remove stake Entry after decreasing provider self delegation", err,
+				utils.Attribute{Key: "index", Value: index},
+				utils.Attribute{Key: "spec", Value: chainID},
+			)
 		}
+		details["min_self_delegation"] = k.GetParams(ctx).MinSelfDelegation.String()
+		utils.LogLavaEvent(ctx, k.Logger(ctx), types.UnstakeFromUnbond, details, "unstaking provider due to unbond that lowered its stake below min self delegation")
+		unstakeHoldBlocks := k.epochstorageKeeper.GetUnstakeHoldBlocks(ctx, stakeEntry.Chain)
+		return k.epochstorageKeeper.AppendUnstakeEntry(ctx, stakeEntry, unstakeHoldBlocks)
+	} else if stakeEntry.EffectiveStake().LT(k.specKeeper.GetMinStake(ctx, chainID).Amount) {
+		details["min_spec_stake"] = k.specKeeper.GetMinStake(ctx, chainID).String()
+		utils.LogLavaEvent(ctx, k.Logger(ctx), types.FreezeFromUnbond, details, "freezing provider due to unbond that lowered its stake below min spec stake")
```

Comment:
not necessarily unbond, can be initial stake is too low

Reply:
done